### ISSUE TITLE
Consider how to make use of containerization

### ISSUE THREAD

**Issue by Jan-Willem** [casangi/cloudviper#3]

**Comment by amcnicho** [comment]
In the short term we plan to develop following the model of the CNGI prototype benchmarks: modifying containers from [open source application deployments ](https://medium.com/pangeo/pangeo-cloud-cluster-design-9d58a1bf1ad3)(provided via helm charts) to add our own runtime dependencies. This approach should be fine to prove concepts but I expect it will not satisfy the requirements of production-grade operations, as we will want to build our own deployment, including new containers.

Ultimately if we want our [applications](https://12factor.net/) to scale using the [Kubernetes API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/) then it’s probably better to build and deploy [our own images](https://kubernetes.io/docs/concepts/containers/images/#use-cases). We should be composing [workloads](https://kubernetes.io/docs/concepts/workloads/) out of our domain specific functions (e.g., [front-end](https://github.com/casangi/casagui/blob/main/casagui/apps/_interactiveclean.py#L1131), [back-end](https://github.com/casangi/astroviper/tree/main/src/astroviper/_domain), [coordination](https://github.com/casangi/graphviper/tree/main/src/graphviper/graph_tools); whether stateful or stateless, following commonly expressed [design patterns](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/45406.pdf)) assembled into [pods](https://kubernetes.io/docs/concepts/workloads/pods/) running [OCI](https://github.com/opencontainers/runtime-spec/tree/main)-compliant containers, connected by [service](https://kubernetes.io/docs/concepts/services-networking/service/#services-in-kubernetes) definitions to expose them to users at the [application layer](https://en.wikipedia.org/wiki/OSI_model).

**Comment by whiteheaddmark** [comment]
[Here](https://nrao.sharepoint.com/:w:/s/NRAOngVLA/EWK_8BKwaapJmh5aMeAQOr8BnPfsubEUYAXzi-KsJTANOQ?e=g3x1G0) is an attempt to start a more specific discussion about utilizing containerization.

**Comment by r-xue** [comment]
@amcnicho:

Have we considered using tools (e.g. [localstack](https://github.com/localstack/localstack)) to emulate certain some AWS-components? mostly thinking from an easy dev/testing perspective. We still don't have a K8s cluster setup on the premises to experiment with things?

sorry.. I might be missing the big picture here: is the goal here to demonstrate "we can do it in the AWS", or how much performance we can get in the actual cloud environment? `cloud` is the in the repo name, but would we consider running locally / on-premises as options just for testing purposes. Maybe it's out of the scope of this repo?

**Comment by amcnicho** [comment]
As I understand it the scope of this repo was at first to track the configuration of the deployment on commercial platforms, and then to evolve it along with the architectural choices as to how that deployment will scale.

The starting point of the configs that became this repo were experiments to both demonstrate we can run the same stack in a commercial cloud ecosystem and compare performance to our traditional HPC approach. Interfacing with the service providers directly was just as much a part of the pathfinding as configuring the deployment.

There were some experiments happening in the systems group (#2) to spin up a cluster partition for testing kubernetes on premises (I'm not sure of their status, but IIUC that was meant as a learning environment, not a [production grade](https://kubernetes.io/docs/setup/production-environment/#production-considerations) cluster yet). If we eventually coalesce into a "kubernetes oriented architecture" then having some of the AWS surface mocked in house to test against would certainly seem helpful in some context, but more concretely defining our own services seems like it will be more fruitful at this point.

**Comment by whiteheaddmark** [comment]
I agree with Andrew - we're setting things up to evaluate horizontal scalability and reliability for on-premises HPC and the cloud. I think there are K8s clusters for evaluation in Charlottesville and Socorro. I was thinking of starting to define services after the upcoming requirements review but I'm open to thinking about that sooner.

**Comment by dklyons81** [comment]
Two points.

1) In conversation with Anthony Johnston last week, he indicated that there is no K8s cluster available in CV, only a certain machine with a partial installation on it. Similarly, in DSOC I have two machines I am using to learn and test with, but there is no test cluster available for development AFAIK. There certainly isn't anything like a supported K8s cluster within NRAO. For K8s testing, I think we are mostly using minikube locally.

2) I have to respond to "having some of the AWS surface mocked in house to test against." I think AWS is sort of the wrong place to start. Rafael's calculations indicate that even a classic HPC setup may have trouble keeping up with data rates that we need to be concerned with. Astra joined us from Google and assumed we would go for a GCloud or AWS approach to ngDP; after about two months she had changed her mind completely and was pushing for building a large data center. This is because there just won't be the bandwidth between us and AWS or any other hosted platform for it to be our _primary_ solution.

I believe cloud-based or hosted services belong in the solution set for scaling beyond what we can do in the data center (demand bursts, perhaps) and for bring-your-own-compute. For that reason, I would rather see a narrow interface to compute that is easy for both AWS and NRAO to adhere or be adapted to, rather than saying one will need to emulate the other. For instance, it would be better to say "we need object storage" and why, and then say "AWS provides S3, compatible providers and/or the internal data center can provide MinIO or other S3-compatible technology." It would be best to say we need object storage for reasons X and Y, rather than we need AWS thus we have to have S3. The analysis can be based on object storage rather than AWS performance specifically.

I haven't pressed the point very hard because I didn't feel like there was much danger of baking AWS into the design, and I still don't, because I trust you guys, but the wording above made me feel like I should chime in on this one.

**Comment by whiteheaddmark** [comment]
Jeff seems to think we have a two-node K8s cluster set up in CV. I'm tracking that down now...

From Jeff, Matthew and Anthony:
"I asked Anthony, who is leading the Kubernetes efforts for my team. He says that the test cluster which I had created in CV is no longer operational. So apparently we took apart the one we had and haven’t put it back together again. Seems like a capability that needs to go into the first PI."

I have asked for this capability to be restored - the intent is a simple production-like cluster that is perhaps slightly more realistic than running minikube locally.
